rfd/0207-encrypted-session-recordings.md

```diff
@@ -0,0 +1,284 @@
+using identities associated with any recipient used during encryption, it is possible
+to verify that new recordings are still retrievable before removing the rotated
+recipient. Reencrypting recordings captured before the rotation is not in scope for
+this implementation. A potential method for ensuring continued access to historical
+recordings would be to generate both a short term and long term recipient. The short
+term recipient would be used for typical decryption operations and playback but
+would be subject to rotations. The long term recipient should only ever be used in
+recovery cases where a session's short term recipients have been rotated.
+
+### Security
+
+The approach described avoids any secret key management or transport within Teleport
+managed infrastructure. All interactions with private keys, either manually provided
+or through integration with an HSM, happen through the client tool `tsh`. This should
+remove most concerns around proper handling of secret keys outside of ensuring that
+`tsh` does not output or transmit any secret keys it has access to.
+
+One of the primary concerns outside of key management is ensuring that session
+recording data is always encrypted before landing on disk or in long term storage. In
+order to help enforce this, all session recording interactions should be gated behind
+a standard interface that can be implemented as either plaintext or encrypted. This
+will help ensure that once the encrypted writer has been selected, any interactions
+with session recordings are encrypted by default.
+
+## UX Examples
+### Teleport admin generating an X25519 key pair
+```bash
+tsh keygen
+private:
+AGE-SECRET-KEY-1L0QNJCXNE546SM9AZYUDX2TQP90GJETG8AKYC8Q9J235K44TRX5SNEJW9K
+public:
+age18l5tzttm32yqvh9pw3dlwp8h4qlvqnx8a40hf7svyag8hnumlsus43gc0y
+```
+
+### Teleport admin wants generating a recipient for an AWS CloudHSM wrapped key
+```bash
+tsh keygen --hsm=aws
+public:
+age18l5tzttm32yqvh9pw3dlwp8h4qlvqnx8a40hf7svyag8hnumlsus43gc0yYWdlLWVuY3J5cHRpb24ub3JnL3YxCi0+IFgyNTUxOSA3V0YyR3laWFNXcHFTMXEz
+```
+
+### Teleport admin wants generating a recipient for a PKCS#11 HSM wrapped key
+```bash
+tsh keygen --hsm=pkcs11 --hsm-pin=12345
+public:
+age18l5tzttm32yqvh9pw3dlwp8h4qlvqnx8a40hf7svyag8hnumlsus43gc0yYWdlLWVuY3J5cHRpb24ub3JnL3YxCi0+IFgyNTUxOSA3V0YyR3laWFNXcHFTMXEz
+```
+
+### Teleport admin wants to replay with multiple private keys
+```yaml
+# teleport.yaml
+auth_service:
+  session_recording_recipients:
+  - recipient_type: age_x25519
+    public_key: age18l5tzttm32yqvh9pw3dlwp8h4qlvqnx8a40hf7svyag8hnumlsus43gc0y
+  - recipient_type: ssh_ed25519
+    public_key: ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAdHXsIFUdyPPNrOQjhuB5SCmGfxSoMSXVZnarhzXL+I erik@Eriks-MacBook-Pro.local
+```
+
+### Teleport admin replaying encrypted session with a given private key
+```bash
+cat ~/my-key | tsh play --identity-stdin 49608fad-7fe3-44a7-b3b5-fab0e0bd34d1
```

Comment:
Nit: `identity-stdin` sounds like _user identity_, not like something necessary for decryption.

Reply:
Maybe `--private-key-stdin` instead? The "identity" was just in keeping with how `age` refers to things, but in retrospect I agree it's way less obvious

Reply:
Sure, or `--decryption-key-stdin` to be even more clear.